### PR TITLE
Nodes attached to the cluster without MCM support are now annotated

### LINKED ISSUE #598: Nodes created with non matching machine class tags aren't orphan collected
**What would you like to be added**:

Add an annotation/label that the MCM does not manage the node. This way higher-level controllers can filter out those nodes during various checks.

**Why is this needed**:

The logic for orphan VM deletion is currently based on tags. We try to filter out all VMs matching with certain tags on the VM (cluster tag and role tag). And if/when we find a VM not matching this tag, we treat it as an orphan and delete it. 

The other way to orphan collect would be to delete any node joining the cluster that doesn't have a matching machine object. However, that doesn't cover cases where we have created VMs but haven't joined as nodes. The secrets are fetched based on machine class and for an orphan VM, we won't know which secret to try and would have try using all machineClass secrets for a cluster. Hence we have ignored this orphan logic in the past.

However, there are cases where we would like to test a new VM type creation by purposefully creating a test VM to join the cluster independent of MCM involvement. And currently, such VMs aren't tagged or marked. 

**Anything else to know**

The idea would be to update [these informer events](https://github.com/gardener/machine-controller-manager/blob/master/pkg/controller/controller.go#L350-L354) to handle nodes without backing machine object and annotate them once and then ignore these nodes for reconcilation in future.

/area robustness
/priority 2

### PR DESCRIPTION
**Pull request by himanshu-kun**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #598 

**Special notes for your reviewer**:
unit tests for helper function `updateNodeWithAnnotations` are also added
triggered safety controller after annotating node object so that if the VM is to be orphan collected it gets done and we don't have to wait for 30min in worst case

**Release note**:

```other operator
Nodes attached to the cluster without MCM support are now annotated with `"node.machine.sapcloud.io/notManagedByMCM": "1"`. This is then ignored by the MCM for further processing.
```

### CODE REVIEW

**Comment by gardener-robot-ci-3** [comment]
Thank you @himanshu-kun for your contribution. Before I can start building your PR, a member of the organization must set the required label(s) {'reviewed/ok-to-test'}. Once started, you can check the build status in the PR checks section below.

File: pkg/util/provider/machinecontroller/machine.go [hunk 1]
```diff
@@ -223,10 +236,13 @@ func (c *controller) deleteNodeToMachine(obj interface{}) {
 
 	machine, err := c.getMachineFromNode(key)
 	if err != nil {
-		klog.Errorf("Couldn't fetch machine %s, Error: %s", key, err)
-		return
-	} else if machine == nil {
-		return
+		if err.Error() == "Multiple machines matching node" {
```
Comment (AxiomSamarth) [inline pkg/util/provider/machinecontroller/machine.go:239]: Why not use `multipleMachineMatchError` like above?

File: pkg/util/provider/machinecontroller/machine.go [hunk 2]
```diff
@@ -223,10 +236,13 @@ func (c *controller) deleteNodeToMachine(obj interface{}) {
...
+			klog.Errorf("Couldn't fetch machine %s, Error: %s", key, err)
+			return
+		} else if err.Error() == "no machines matching node found" {
```
Comment (AxiomSamarth) [inline pkg/util/provider/machinecontroller/machine.go:242]: Same. Why not use `noMachineMatchError` like above?

File: pkg/util/provider/machinecontroller/machine_util.go [hunk 3]
```diff
@@ -1300,3 +1299,23 @@ func getProviderID(machine *v1alpha1.Machine) string {
...
+
+func (c *controller) patchNodeWithAnnotation(node *v1.Node, annotationKey string) {
+
+	// Initialize node annotations if empty
+	if node.Annotations == nil {
+		node.Annotations = make(map[string]string)
+	}
+
+	//append annotation
+	node.Annotations[annotationKey] = "1"
```
Comment (AxiomSamarth) [inline pkg/util/provider/machinecontroller/machine_util.go:1311]: because of this logic, this method may not be reused elsewhere is what I feel. I would like to get advice from other reviewers as well.

**Review by AxiomSamarth — CHANGES_REQUESTED** [review]

File: pkg/util/provider/machinecontroller/machine.go (hunk 1 above)
Comment (himanshu-kun) [reply pkg/util/provider/machinecontroller/machine.go:239]: Oh yes I forgot to make the change here I guess

**Review by himanshu-kun — COMMENTED** [review]
the files I have made changes to are : 
`machineutils/utils.go`
`machine.go`
`machine_util.go`
`machine_util_test.go`
`machine_test.go`

Sorry for the inconvenience, I will make sure to make the changed files as seperate commits hereafter

**Comment by himanshu-kun** [comment]
I will raise another PR with major changes suggested by Amshu

**Review by himanshu-kun — COMMENTED** [review]
the files I have made changes to are :
machineutils/utils.go
machine.go
machine_test.go
machine_util.go
machine_util_test.go
machine_safety.go
machine_safety_util.go
`machine_safety_util.go`
`machine_safety_util_test.go`

(The last two highlighted files here are new in the repo, added by me)

Sorry for the inconvenience, I will make sure to make the changed files as seperate commits hereafter

**Review by AxiomSamarth — CHANGES_REQUESTED** [review]

File: pkg/util/provider/machinecontroller/machine_safety_util.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
+package controller
+
+import (
+	"github.com/gardener/machine-controller-manager/pkg/util/provider/machineutils"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/klog"
+)
+
+func (c *controller) patchNodeWithAnnotation(node *v1.Node, annotationKey string) (machineutils.RetryPeriod, error) {
```
[the remark this answers is not in the document]
Comment (himanshu-kun) [reply pkg/util/provider/machinecontroller/machine_safety_util.go]: solved in commit 7941be4450b95d9b3c6a7b708e9b123c307c8559

File: pkg/util/provider/machinecontroller/machine_safety_util.go [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+	// Initialize node annotations if empty
+	if node.Annotations == nil {
+		node.Annotations = make(map[string]string)
+	}
+
+	//append annotation
+	node.Annotations[annotationKey] = "1"
```
[the remark this answers is not in the document]
Comment (himanshu-kun) [reply pkg/util/provider/machinecontroller/machine_safety_util.go]: solved in commit 7941be4450b95d9b3c6a7b708e9b123c307c8559

File: pkg/util/provider/machinecontroller/machine_util.go [hunk 6]
```diff
@@ -42,7 +42,6 @@ import (
 	utilstrings "github.com/gardener/machine-controller-manager/pkg/util/strings"
 	utiltime "github.com/gardener/machine-controller-manager/pkg/util/time"
 
-	corev1 "k8s.io/api/core/v1"
```
Comment (prashanth26) [inline pkg/util/provider/machinecontroller/machine_util.go:45]: Nice catch. Thanks for the this change. 

Ideally, we keep `corev1` and remove `v1` as v1 could refer to v1 of many different libraries. CoreV1 gives more context.  Anyways should be okay for now.

**Review by prashanth26 — COMMENTED** [review]
Thanks for the changes. PR overall looks good. Just minor NITs. 

Also can you split the PR with 2 commits one for actual changes and other for revedoring. Also please update the release notes. Thanks.

File: pkg/util/provider/machinecontroller/machine_safety_util.go [hunk 7]
```diff
@@ -0,0 +1,29 @@
+package controller
+
+import (
+	"github.com/gardener/machine-controller-manager/pkg/util/provider/machineutils"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/klog"
+)
+
+func (c *controller) patchNodeWithAnnotation(node *v1.Node, annotations map[string]string) (machineutils.RetryPeriod, error) {
+
```
[the remark this answers is not in the document]
Comment (himanshu-kun) [reply pkg/util/provider/machinecontroller/machine_safety_util.go:10]: I guess not @prashanth26  . I am modifying a copy of node object which I fetched from cache and then updating the object at APIServer(means at etcd) not the one in cache.

File: pkg/util/provider/machineutils/utils.go [hunk 8]
```diff
@@ -47,6 +47,9 @@ const (
 	// MigratedMachineClass annotation helps in identifying machineClasses who have been migrated by migration controller
 	MigratedMachineClass = "machine.sapcloud.io/migrated"
 
+	// NotHanldedByMCM annotation helps in identifying the nodes which are not handled by MCM
+	NotHandledByMCM = "node.machine.sapcloud.io/notHanldedByMCM"
```
[the remark this answers is not in the document]
Comment (himanshu-kun) [reply pkg/util/provider/machineutils/utils.go:51]: thanks for pointing out the typo :0

[the remark this answers is not in the document]
Comment (himanshu-kun) [reply pkg/util/provider/machineutils/utils.go:51]: yes 'managed' is better

**Comment by himanshu-kun** [comment]
@prashanth26 pls let me know if release not is alright.

File: pkg/util/provider/machinecontroller/machine_safety.go [hunk 9]
```diff
@@ -161,6 +168,39 @@ func (c *controller) isAPIServerUp() bool {
...
+	ClusterNodes, err := c.nodeLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("Safety-Net: Error getting nodes")
+		return machineutils.LongRetry, err
+	}
+	//getMachineFromNode()
+	for _, node := range ClusterNodes {
+		_, err := c.getMachineFromNode(node.Name)
+		if err != nil {
+			if err.Error() == multipleMachineMatchError {
```
Comment (amshuman-kr) [inline pkg/util/provider/machinecontroller/machine_safety.go:183]:
```suggestion
			if err == multipleMachineMatchError {
```

File: pkg/util/provider/machinecontroller/machine_safety.go [hunk 10]
```diff
@@ -161,6 +168,39 @@ func (c *controller) isAPIServerUp() bool {
...
+				klog.Errorf("Couldn't fetch machine, Error: %s", err)
+			} else if err.Error() == noMachineMatchError {
```
Comment (amshuman-kr) [inline pkg/util/provider/machinecontroller/machine_safety.go:185]:
```suggestion
			} else if err == noMachineMatchError {
```

File: pkg/util/provider/machinecontroller/machine_safety_util.go [hunk 11]
```diff
@@ -0,0 +1,29 @@
+package controller
+
+import (
+	"github.com/gardener/machine-controller-manager/pkg/util/provider/machineutils"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/klog"
+)
+
+func (c *controller) updateNodeWithAnnotation(node *v1.Node, annotations map[string]string) (machineutils.RetryPeriod, error) {
```
Comment (amshuman-kr) [inline pkg/util/provider/machinecontroller/machine_safety_util.go:9]: No need for this func to return the retry period. The caller should decide the retry period (which is a constant here anyway). Does it matter if it is medium or long retry?

File: pkg/util/provider/machinecontroller/machine_test.go [hunk 12]
```diff
@@ -2794,7 +2794,6 @@ var _ = Describe("machine", func() {
 			}),
 		)
 	})
-
```
Comment (amshuman-kr) [inline pkg/util/provider/machinecontroller/machine_test.go:2797]: Is this change required?

File: pkg/util/provider/machineutils/utils.go [hunk 13]
```diff
@@ -47,6 +47,9 @@ const (
 	// MigratedMachineClass annotation helps in identifying machineClasses who have been migrated by migration controller
 	MigratedMachineClass = "machine.sapcloud.io/migrated"
 
+	// NotHanldedByMCM annotation helps in identifying the nodes which are not handled by MCM
+	NotManagedByMCM = "node.machine.sapcloud.io/notManagedByMCM"
```
Comment (amshuman-kr) [inline pkg/util/provider/machineutils/utils.go:51]:
```suggestion
	NotManagedByMCM = "node.machine.sapcloud.io/not-managed-by-mcm"
```
_CamelCase_ may not be suitable for annotations.

**Review by amshuman-kr — CHANGES_REQUESTED** [review]
Thanks for addressing my comments @himanshu-kun! Some comments below.

Also wouldn't an annotation like `node.machine.sapcloud.io/not-managed-by-mcm: "true"` be better for convention?

File: pkg/util/provider/machinecontroller/machine_safety.go [hunk 14]
```diff
@@ -161,6 +168,39 @@ func (c *controller) isAPIServerUp() bool {
...
+//AnnotateNodesUnmanagedByMCM checks for nodes which are not handled by MCM and annotes them
+func (c *controller) AnnotateNodesUnmanagedByMCM() (machineutils.RetryPeriod, error) {
+	//list all the nodes on target cluster
+	ClusterNodes, err := c.nodeLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("Safety-Net: Error getting nodes")
+		return machineutils.LongRetry, err
+	}
+	//getMachineFromNode()
+	for _, node := range ClusterNodes {
```
[the remark this answers is not in the document]
Comment (himanshu-kun) [reply pkg/util/provider/machinecontroller/machine_safety.go:180]: I have made the change, but isn't `nodes` an array of pointers to node structs?

File: pkg/util/provider/machineutils/utils.go (hunk 13 above)
Comment (himanshu-kun) [reply pkg/util/provider/machineutils/utils.go:51]: Noted, thanks ..will take care from next time.

File: pkg/util/provider/machinecontroller/machine_test.go (hunk 12 above)
Comment (himanshu-kun) [reply pkg/util/provider/machinecontroller/machine_test.go:2797]: No. I guess I might have missed that this is a small change and added the file to commit.

File: pkg/util/provider/machinecontroller/machine_safety_util.go (hunk 11 above)
Comment (himanshu-kun) [reply pkg/util/provider/machinecontroller/machine_safety_util.go:9]: @amshuman-kr MediumRetry is 2 min, LongRetry is 10 min. I am returning MediumRetry when I'm unable to patch the node object on API Server . A MediumRetry looked more time effective in this case , so I have used it. 
If it doesn't make much difference , then I can remove it.

File: pkg/util/provider/machinecontroller/machine.go [hunk 15]
```diff
@@ -198,10 +207,12 @@ func (c *controller) addNodeToMachine(obj interface{}) {
...
 	machine, err := c.getMachineFromNode(key)
 	if err != nil {
-		klog.Errorf("Couldn't fetch machine %s, Error: %s", key, err)
-		return
-	} else if machine == nil {
-		return
+		if err.Error() == multipleMachineMatchError {
+			klog.Errorf("Couldn't fetch machine %s, Error: %s", key, err)
+			return
+		} else if err.Error() == noMachineMatchError {
```
[the remark this answers is not in the document]
Comment (himanshu-kun) [reply pkg/util/provider/machinecontroller/machine.go:213]: sorry I didn't get you Amshu, can you elaborate.

File: pkg/util/provider/machinecontroller/machine_safety.go [hunk 16]
```diff
@@ -161,6 +168,39 @@ func (c *controller) isAPIServerUp() bool {
 	return true
 }
 
+//AnnotateNodesUnmanagedByMCM checks for nodes which are not handled by MCM and annotes them
+func (c *controller) AnnotateNodesUnmanagedByMCM() (machineutils.RetryPeriod, error) {
```
[the remark this answers is not in the document]
Comment (himanshu-kun) [reply pkg/util/provider/machinecontroller/machine_safety.go:172]: I did this following how `checkMachineClasses()` was written @amshuman-kr . there also the called function was returning the retry period.

**Review by himanshu-kun — COMMENTED** [review]
@amshuman-kr I have made the changes you asked, they are made in the commit a09aa34

File: pkg/util/provider/machinecontroller/machine.go [hunk 17]
```diff
@@ -198,10 +207,13 @@ func (c *controller) addNodeToMachine(obj interface{}) {
...
-		return
-	} else if machine == nil {
-		return
+		if err == multipleMachineMatchError {
+			klog.Errorf("Couldn't fetch machine %s, Error: %s", key, err)
+			return
+		} else if err == noMachineMatchError {
+			klog.Errorf("Couldn't fetch machine %s, Error: %s", key, err)
+			return
+		}
```
Comment (amshuman-kr) [inline pkg/util/provider/machinecontroller/machine.go:216]: Does this change now make sense given that the handling for the error cases is now pretty much the same?

File: pkg/util/provider/machinecontroller/machine.go [hunk 18]
```diff
@@ -223,10 +235,13 @@ func (c *controller) deleteNodeToMachine(obj interface{}) {
...
```
Comment (amshuman-kr) [inline pkg/util/provider/machinecontroller/machine.go:244]: Does this change now make sense given that the handling for the error cases is now pretty much the same?

File: pkg/util/provider/machinecontroller/machine_safety.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (amshuman-kr) [reply pkg/util/provider/machinecontroller/machine_safety.go:172]: Sorry for suggesting this without the realising that some other called function is also doing it. We can clean this up later separately.

File: pkg/util/provider/machinecontroller/machine_safety.go [hunk 19]
```diff
@@ -161,6 +168,42 @@ func (c *controller) isAPIServerUp() bool {
...
+	//list all the nodes on target cluster
+	nodes, err := c.nodeLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("Safety-Net: Error getting nodes")
+		return machineutils.LongRetry, err
+	}
+	//getMachineFromNode()
+	for i := range nodes {
+		node := nodes[i]
```
Comment (amshuman-kr) [inline pkg/util/provider/machinecontroller/machine_safety.go:181]:
```suggestion
		node := &nodes[i]
```
Otherwise it would still be a shallow-copy.

**Review by amshuman-kr — COMMENTED** [review]
Thanks for changes @himanshu-kun! A couple of nit-picks. LGTM otherwise.

File: pkg/util/provider/machinecontroller/machine.go (hunk 18 above)
Comment (himanshu-kun) [reply pkg/util/provider/machinecontroller/machine.go]: @amshuman-kr  review addressed in b6a183c..brought back the discussed section of code to the old state.. please recommend code in case you feel any changes needed.

**Review by amshuman-kr — APPROVED** [review]
Thanks! /lgtm

File: pkg/util/provider/machinecontroller/machine_safety.go [hunk 20]
```diff
@@ -161,6 +168,41 @@ func (c *controller) isAPIServerUp() bool {
...
 
+//AnnotateNodesUnmanagedByMCM checks for nodes which are not handled by MCM and annotes them
+func (c *controller) AnnotateNodesUnmanagedByMCM() (machineutils.RetryPeriod, error) {
+	//list all the nodes on target cluster
+	nodes, err := c.nodeLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("Safety-Net: Error getting nodes")
+		return machineutils.LongRetry, err
+	}
+	//getMachineFromNode()
```
Comment (prashanth26) [inline pkg/util/provider/machinecontroller/machine_safety.go:179]:
```suggestion
```

File: pkg/util/provider/machinecontroller/machine_safety.go [hunk 21]
```diff
@@ -161,6 +168,41 @@ func (c *controller) isAPIServerUp() bool {
 	return true
 }
 
+//AnnotateNodesUnmanagedByMCM checks for nodes which are not handled by MCM and annotes them
```
Comment (prashanth26) [inline pkg/util/provider/machinecontroller/machine_safety.go:171]:
```suggestion
// AnnotateNodesUnmanagedByMCM checks for nodes which are not handled by MCM and annotes them
```

File: pkg/util/provider/machineutils/utils.go [hunk 22]
```diff
@@ -47,6 +47,9 @@ const (
 	// MigratedMachineClass annotation helps in identifying machineClasses who have been migrated by migration controller
 	MigratedMachineClass = "machine.sapcloud.io/migrated"
 
+	// NotHanldedByMCM annotation helps in identifying the nodes which are not handled by MCM
```
Comment (prashanth26) [inline pkg/util/provider/machineutils/utils.go:50]:
```suggestion
	// NotManagedByMCM annotation helps in identifying the nodes which are not handled by MCM
```

File: pkg/util/provider/machinecontroller/machine.go [hunk 23]
```diff
@@ -190,6 +195,10 @@ func (c *controller) addNodeToMachine(obj interface{}) {
 		return
 	}
 
+	if _, annotationPresent := node.ObjectMeta.Annotations[machineutils.NotManagedByMCM]; annotationPresent {
```
Comment (prashanth26) [inline pkg/util/provider/machinecontroller/machine.go:198]:
```suggestion
	 // If NotManagedByMCM annotation is present on node, don't process this node object
	if _, annotationPresent := node.ObjectMeta.Annotations[machineutils.NotManagedByMCM]; annotationPresent {
```
Just adding a comment for clarity for external contributors.

**Review by prashanth26 — COMMENTED** [review]
Just minor NITS. You can add all suggestioned changes to a branch from PR review files and make a commit of it. 

- Finally would be great to split the PR into two commits - Vendoring changes and actual changes. 
- Keep meaningful commit messages
- Capitalize first word in commit & PR messages for consistency. 

Thanks.

**Comment by prashanth26** [comment]
/title Nodes attached to the cluster without MCM support are now annotated

**Review by prashanth26 — APPROVED** [review]
/lgtm

**Review by AxiomSamarth — APPROVED** [review]
